### PR TITLE
Added git plumbing tests to check the OTP repository state

### PR DESCRIPTION
**Pull request by ziopio**
These functions allow to easily add git commands and predicates to evaluate the git outputs.
Right now i just check that fetch returns nothing and the OTP tag is correct.

If something is wrong the directory is deleted and the repo cloned again.

### CODE REVIEW

**Comment by ziopio** [comment]
[Check for Git status when verifying checkout](https://trello.com/c/gDbmsNW4/4-check-for-git-status-when-verifying-checkout)

File: src/grisp_tools_build.erl [hunk 1]
```diff
@@ -63,24 +63,23 @@ build(S0) ->
 
 download(#{otp_version := {_,_,_,Ver}} = S0) ->
     BuildPath = mapz:deep_get([paths, build], S0),
-    % FIXME: Check actual state of checkout instead of just checking for folder
-    % presence!
-    case filelib:is_dir(filename:join(BuildPath, ".git")) of
-        false ->
+    case check_git_repository(BuildPath, S0) of
+        {false, S1}->
```
Comment (eproxus) [inline src/grisp_tools_build.erl]: It is a bit hard to follow what `true` and `false` means in these contexts. We should use dedicated atoms to convey meaning instead:

```erl
case check_git_repository(BuildPath, S0) of
    {up_to_date, S1} ->
        event(mapz:deep_put([build, download], false, S1), ['_skip']);
    {_Other, S1} ->
        ...
    end.
```

**Comment by eproxus** [comment]
An idea I just had: would the code be clearer if we separated the check and download as two steps? I.e. something like:

```
    grisp_tools_util:weave(State, [
        fun grisp_tools_step:config/1,
        {fun build/1, [
            ...,
            {repo, [
                fun check/1,
                fun clone/1
            ]}
            ...
...


check_repo(S0) ->
    {State, S1} = check_git_repository(BuildPath, S0)
    S1#{repo_state => State}.

clone_repo(#{repo_state := up_to_date} = S0) ->
    event(mapz:deep_put([build, download], false, S1), ['_skip']);
clone_repo(#{repo_state := State} = S0) ->
    S1 = event(S0, [clone, State]);
    BuildPath = mapz:deep_get([paths, build], S1),
    S2 = delete_directory(BuildPath, S1),
    ...
```

That would get rid of many nested case statements.

**Comment by ziopio** [comment]
Ok i applied your suggestions.

I don't fire new events, for them i should add lines in rebar3_grisp...

**Comment by ziopio** [comment]
Last commit fixes this type of error:
```
===> Building OTP for GRiSP
* Platform: grisp2
* Resolving OTP version
    23.3.4.11 (requirement was "23.3")
* Downloading
===> sh(git clone -b OTP-23.3.4.11 --depth 1 --single-branch https://github.com/erlang/otp /mnt/c/Users/Luca Succi/Desktop/grisp_demo/_grisp/grisp2/otp/23.3.4.11/build)
failed with return code 129 and the following output:
fatal: Too many arguments.
```
But this does not solve completly the problem since goes out of the scope of this plugin.
Indeed with a spaced directory it crashes later at the config stage
```
/mnt/c/Users/Luca Succi/Desktop/grisp_demo/_grisp/grisp2/otp/23.3.4.11/build/configure  --cache-file=/dev/null --enable-bootstrap-only
=== Running configure in /mnt/c/Users/Luca Succi/Desktop/grisp_demo/_grisp/grisp2/otp/23.3.4.11/build/erts ===
./configure --enable-bootstrap-only --disable-option-checking --cache-file=/dev/null --srcdir="/mnt/c/Users/Luca Succi/Desktop/grisp_demo/_grisp/grisp2/otp/23.3.4.11/build/erts"
./configure: line 2902: cd: too many arguments
```

File: src/grisp_tools_build.erl [hunk 2]
```diff
@@ -281,6 +285,12 @@ run_hooks(S0, Type, Opts) ->
             end, S1, lists:sort(maps:to_list(Hooks)))
     end.
 
+delete_directory(BuildPath, S0) ->
+    case filelib:is_dir(BuildPath) of
+        true -> shell_ok(S0, "rm -rf "++BuildPath, []);
```
Comment (eproxus) [inline src/grisp_tools_build.erl]: If we are fixing spaces in paths, we should add quotes here too.

**Review by eproxus — APPROVED** [review]
